### ISSUE TITLE
SSL Error When Attempting To Install Pip Packages During Build

### ISSUE THREAD

**Issue by jsikarin** [ansible/ansible-builder#573]
I'm trying to build a container containing the "theforeman.foreman" collection. My current build results in the following error.

```
sudo /usr/local/bin/ansible-builder build --tag my_first_ee_image
[sudo] password for user:
Running command:
  podman build -f context/Containerfile -t my_first_ee_image context
...showing last 20 lines of output...
--> Using cache f59a3d23eb80be59365d67f1d902ada0abdc1c0bb61c13c945d76f12281e57f4
--> f59a3d23eb8
[3/4] STEP 7/12: ARG ANSIBLE_GALAXY_CLI_COLLECTION_OPTS
--> Using cache 007ef3ccda9d1948327dbe65933a606b22c5a175fd2c4020920dfec35aaf0789
--> 007ef3ccda9
[3/4] STEP 8/12: ARG ANSIBLE_GALAXY_CLI_ROLE_OPTS
--> Using cache 41607f237acf2832de5988c927808ef7f2c7106a72483e2aa105f9c979568fe0
--> 41607f237ac
[3/4] STEP 9/12: RUN $PYCMD -m pip install --no-cache-dir bindep pyyaml requirements-parser
Collecting bindep
  WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1131)'))': /packages/f8/94/5a1059b364ae3534e941950f21fbdceddef20e1aa4d17f373c375bcc1883/bindep-2.11.0-py2.py3-none-any.whl
  WARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1131)'))': /packages/f8/94/5a1059b364ae3534e941950f21fbdceddef20e1aa4d17f373c375bcc1883/bindep-2.11.0-py2.py3-none-any.whl
  WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1131)'))': /packages/f8/94/5a1059b364ae3534e941950f21fbdceddef20e1aa4d17f373c375bcc1883/bindep-2.11.0-py2.py3-none-any.whl
  WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1131)'))': /packages/f8/94/5a1059b364ae3534e941950f21fbdceddef20e1aa4d17f373c375bcc1883/bindep-2.11.0-py2.py3-none-any.whl
  WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1131)'))': /packages/f8/94/5a1059b364ae3534e941950f21fbdceddef20e1aa4d17f373c375bcc1883/bindep-2.11.0-py2.py3-none-any.whl
ERROR: Could not install packages due to an OSError: HTTPSConnectionPool(host='files.pythonhosted.org', port=443): Max retries exceeded with url: /packages/f8/94/5a1059b364ae3534e941950f21fbdceddef20e1aa4d17f373c375bcc1883/bindep-2.11.0-py2.py3-none-any.whl (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1131)')))

WARNING: You are using pip version 22.0.4; however, version 23.1.2 is available.
You should consider upgrading via the '/usr/bin/python3 -m pip install --upgrade pip' command.
Error: building at STEP "RUN $PYCMD -m pip install --no-cache-dir bindep pyyaml requirements-parser": while running runtime: exit status 1

An error occurred (rc=1), see output line(s) above for details.
```

Here are the contents of my build file

```
---
version: 1
dependencies:
  galaxy: requirements.yml
build_arg_defaults:
  ANSIBLE_GALAXY_CLI_COLLECTION_OPTS: '--ignore-certs'
```

Please note I added 

build_arg_defaults:
  ANSIBLE_GALAXY_CLI_COLLECTION_OPTS: '--ignore-certs'

to deal with a previous error downloading the collections from ansible galaxy. So this is the second error I'm trying to solve.

**Comment by Akasurde** [comment]
@jsikarin Thanks for reporting this issue. Could you please try `version 3` of ee definition and let us know if it works for you? 

example ee looks like - 

```yaml
---
version: 3

dependencies:
  galaxy: requirements.yml
```

and you will need to install ansible-builder 3.

**Comment by jsikarin** [comment]
@Akasurde 
that brings up a new error as show below 

```
sudo /usr/local/bin/ansible-builder build --tag my_first_ee_image
Running command:
  podman build -f context/Containerfile -t my_first_ee_image context
...showing last 20 lines of output...
Requirement already satisfied: setuptools in /usr/lib/python3.8/site-packages (41.6.0)
Requirement already satisfied: pip in /usr/local/lib/python3.8/site-packages (22.0.4)
+ '[' -z /usr/bin/dnf ']'
+ '[' /usr/bin/dnf = /usr/bin/microdnf ']'
+ '[' -f /output/bindep/run.txt ']'
++ cat /output/bindep/run.txt
+ PACKAGES='python3-rpm
python38-requests'
+ '[' '!' -z 'python3-rpm
python38-requests' ']'
+ /usr/bin/dnf install -y python3-rpm python38-requests
CentOS Stream 8 - AppStream                     0.0  B/s |   0  B     00:01
Errors during downloading metadata for repository 'appstream':
  - Curl error (56): Failure when receiving data from the peer for http://mirrorlist.centos.org/?release=8-stream&arch=x86_64&repo=AppStream&infra=stock [Recv failure: Connection reset by peer]
Error: Failed to download metadata for repo 'appstream': Cannot prepare internal mirrorlist: Curl error (56): Failure when receiving data from the peer for http://mirrorlist.centos.org/?release=8-stream&arch=x86_64&repo=AppStream&infra=stock [Recv failure: Connection reset by peer]
Error: building at STEP "RUN /output/scripts/install-from-bindep && rm -rf /output/wheels": while running runtime: exit status 1

An error occurred (rc=1), see output line(s) above for details.
```

I have not yet found a way to change the repo so that I can install packages such as python3-rpm and python38-requests into my container. If you know how I could fix this error I would appreciate the help. Thanks for pointing out I can change the version number. I never would have through that would do anything

**Comment by budgester** [comment]
Looks like you maybe behind a proxy, I've had to use our internal pypi proxy (jfrog/nexus) by copy a custom pip.conf file onto the server.

**Comment by jsikarin** [comment]
okay at the moment I'm not sure if my company has a proxy I can use(but I will find out) so I have a couple questions

1. Is there a way I can specify a CA in either the build file, ansible.cfg, or pip.conf?
2. Why does the build file default to CentOS 8 Stream? What if I wanted to use a different set of repos like Redhat. Could I do that? If so how?

**Comment by Akasurde** [comment]
@jsikarin 
1. Check https://ansible.readthedocs.io/projects/builder/en/latest/scenario_guides/scenario_copy/#copying-arbitratory-files-to-ee to specify a custom CA certificate.
2. You can check `images` to specify different container images - https://ansible.readthedocs.io/projects/builder/en/latest/definition/#images. This section contains a few examples of the images - https://ansible.readthedocs.io/projects/builder/en/latest/definition/#version-3-sample-file

**Comment by jsikarin** [comment]
1. Adding the CA using the referenced link didn't work, which is fine. I don't really want to use the CentOS repos anyway. 
2. I tried using a RHEL 8 image. Here is my current build file

````
---
version: 3
dependencies:
  galaxy: requirements.yml

images:
  base_image:
    name: registry.redhat.io/rhel8/python-38:latest

additional_build_files:
  - src: /usr/share/pki/ca-trust-source/anchors/developer-ca-bundle-20211129.pem
    dest: configs
  - src: /etc/ansible/ansible.cfg
    dest: configs

additional_build_steps:
  prepend_base:
    - COPY _build/configs/rootCA.crt /usr/share/pki/ca-trust-source/anchors
    - RUN update-ca-trust
    - RUN export CURL_CA_BUNDLE=/usr/share/pki/ca-trust-source/anchors/rootCA.crt
    - RUN export REQUESTS_CA_BUNDLE=/usr/share/pki/ca-trust-source/anchors/rootCA.crt
    - RUN python3 -m pip config set global.cert /usr/share/pki/ca-trust-source/anchors/rootCA.crt
  prepend_galaxy:
    - ADD _build/configs/ansible.cfg ~/.ansible.cfg
````

When attempting to build the container with podman I get the following errors.


````
Running command:
  podman build -f context/Containerfile -t my_first_ee_image context
...showing last 20 lines of output...
time="2023-07-07T13:05:09-05:00" level=warning msg="missing \"PYCMD\" build argument. Try adding \"--build-arg PYCMD=<VALUE>\" to the command line"
time="2023-07-07T13:05:09-05:00" level=warning msg="missing \"PKGMGR_PRESERVE_CACHE\" build argument. Try adding \"--build-arg PKGMGR_PRESERVE_CACHE=<VALUE>\" to the command line"
time="2023-07-07T13:05:09-05:00" level=warning msg="missing \"ANSIBLE_GALAXY_CLI_COLLECTION_OPTS\" build argument. Try adding \"--build-arg ANSIBLE_GALAXY_CLI_COLLECTION_OPTS=<VALUE>\" to the command line"
time="2023-07-07T13:05:09-05:00" level=warning msg="missing \"ANSIBLE_GALAXY_CLI_ROLE_OPTS\" build argument. Try adding \"--build-arg ANSIBLE_GALAXY_CLI_ROLE_OPTS=<VALUE>\" to the command line"
time="2023-07-07T13:05:09-05:00" level=warning msg="missing \"PKGMGR\" build argument. Try adding \"--build-arg PKGMGR=<VALUE>\" to the command line"
time="2023-07-07T13:05:09-05:00" level=warning msg="missing \"EE_BASE_IMAGE\" build argument. Try adding \"--build-arg EE_BASE_IMAGE=<VALUE>\" to the command line"
time="2023-07-07T13:05:09-05:00" level=warning msg="missing \"PYCMD\" build argument. Try adding \"--build-arg PYCMD=<VALUE>\" to the command line"
time="2023-07-07T13:05:09-05:00" level=warning msg="missing \"PKGMGR_PRESERVE_CACHE\" build argument. Try adding \"--build-arg PKGMGR_PRESERVE_CACHE=<VALUE>\" to the command line"
time="2023-07-07T13:05:09-05:00" level=warning msg="missing \"ANSIBLE_GALAXY_CLI_COLLECTION_OPTS\" build argument. Try adding \"--build-arg ANSIBLE_GALAXY_CLI_COLLECTION_OPTS=<VALUE>\" to the command line"
time="2023-07-07T13:05:09-05:00" level=warning msg="missing \"ANSIBLE_GALAXY_CLI_ROLE_OPTS\" build argument. Try adding \"--build-arg ANSIBLE_GALAXY_CLI_ROLE_OPTS=<VALUE>\" to the command line"
time="2023-07-07T13:05:09-05:00" level=warning msg="missing \"PKGMGR\" build argument. Try adding \"--build-arg PKGMGR=<VALUE>\" to the command line"
time="2023-07-07T13:05:09-05:00" level=warning msg="missing \"EE_BASE_IMAGE\" build argument. Try adding \"--build-arg EE_BASE_IMAGE=<VALUE>\" to the command line"
time="2023-07-07T13:05:09-05:00" level=warning msg="missing \"PYCMD\" build argument. Try adding \"--build-arg PYCMD=<VALUE>\" to the command line"
time="2023-07-07T13:05:09-05:00" level=warning msg="missing \"PKGMGR_PRESERVE_CACHE\" build argument. Try adding \"--build-arg PKGMGR_PRESERVE_CACHE=<VALUE>\" to the command line"
time="2023-07-07T13:05:09-05:00" level=warning msg="missing \"ANSIBLE_GALAXY_CLI_COLLECTION_OPTS\" build argument. Try adding \"--build-arg ANSIBLE_GALAXY_CLI_COLLECTION_OPTS=<VALUE>\" to the command line"
time="2023-07-07T13:05:09-05:00" level=warning msg="missing \"ANSIBLE_GALAXY_CLI_ROLE_OPTS\" build argument. Try adding \"--build-arg ANSIBLE_GALAXY_CLI_ROLE_OPTS=<VALUE>\" to the command line"
time="2023-07-07T13:05:09-05:00" level=warning msg="missing \"PKGMGR\" build argument. Try adding \"--build-arg PKGMGR=<VALUE>\" to the command line"
[1/4] STEP 1/16: FROM registry.redhat.io/rhel8/python-38:latest AS base
Trying to pull registry.redhat.io/rhel8/python-38:latest...
Error: creating build container: initializing source docker://registry.redhat.io/rhel8/python-38:latest: unable to retrieve auth token: invalid username/password: unauthorized: Please login to the Red Hat Registry using your Customer Portal credentials. Further instructions can be found here: https://access.redhat.com/RegistryAuthentication

An error occurred (rc=125), see output line(s) above for details.
````

Now lets forget about all the vars it says I need supplied for a moment. 
The last error is showing a failure to authenticate to docker://registry.redhat.io. I have added the url, auth_url, and token to my /etc/ansible/ansible.cfg and also tried adding the ansible.cfg to my build file. However I feel the way I am doing this is incorrect because the build file is adding the ansible.cfg file to my container instead of using it for authentication. What is the proper way to add my creds to my execution-environment.yml so that ansible-builder can pull down the container that I want from redhat.registry.io?

**Comment by jsikarin** [comment]
For example this post here: https://gregsowell.com/?p=7086

says to add the line 

`ansible_config: ansible.cfg`

to the build file so that ansible-builder can find the token. This option does not seem to work in v3 of ansible-builder. Is there a different var I can add so the builder can find my token in my ansible.cfg?

**Comment by Akasurde** [comment]
@jsikarin You need to first login to Red Hat container registry before downloading the image from it.
Ansible builder assumes that you are already logged in container registry and does not do it automatically for you.

```
[1/4] STEP 1/16: FROM registry.redhat.io/rhel8/python-38:latest AS base
Trying to pull registry.redhat.io/rhel8/python-38:latest...
Error: creating build container: initializing source docker://registry.redhat.io/rhel8/python-38:latest: unable to retrieve auth token: invalid username/password: unauthorized: Please login to the Red Hat Registry using your Customer Portal credentials. Further instructions can be found here: https://access.redhat.com/RegistryAuthentication

```
In order to get rid of this error, follow the guide under `Using Authentication` at https://access.redhat.com/RegistryAuthentication. Then try to build using EE image.

tl;dr - 

```
# podman login registry.redhat.io
Username: myrhusername
Password: ***********
```
or

```
# docker login registry.redhat.io
Username: myrhusername
Password: ***********
```

**Comment by budgester** [comment]
We copy our own pip.conf file into the container..

```
additional_build_steps:
  prepend_base:
    - RUN $PYCMD -m pip config debug
    - COPY pip.conf /etc/pip.conf
    - RUN $PYCMD -m pip config list
```

**Comment by jsikarin** [comment]
I was able to successfully pull down the image form registry.redhat.io after logging in with podman and build the container with the required ansible collections. Thank you!
